### PR TITLE
refactor: use token instead of constant in size

### PR DESCRIPTION
**Pull request by li-jia-nan**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Fix Space preset `size` not following design token and Compact theme will now correct handle space gap value. |
| 🇨🇳 Chinese | 修复 Space 的预设 `size` 不会跟随 Design Token 的问题，现在紧凑模式也会正确处理对应的间距数值。 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 38d2448</samp>

The pull request refactors and improves the `Space` component and its related files. It uses logical properties and CSS variables to handle different spacing sizes and writing directions, adds custom token support for the `size` prop, and fixes some test cases and code style issues.

### <samp>🤖 Generated by Copilot at 38d2448</samp>

*  Add `SpaceContextType` interface and use it for type annotation in `context.ts` and `Item.tsx` ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-150e4418a7a470f7b98734175d2540bdef98aa1495513ea15a6a1b608b2ff595L3-R10), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a7c0ed1303102439a0949fae071f0b7aa0c4a13694ba55cd513e613c5f3673a2L2-R4), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L124-R122), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a7c0ed1303102439a0949fae071f0b7aa0c4a13694ba55cd513e613c5f3673a2L26-R26))
* Replace `marginRight` and `marginLeft` with `marginInlineStart` in `Item.tsx` and update test cases in `index.test.tsx` to support RTL layouts ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a7c0ed1303102439a0949fae071f0b7aa0c4a13694ba55cd513e613c5f3673a2L37-R37), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-93edd300f3e0d705dfe5647471697e62c1698695f56c65e766221ef0985d859cL69-R72), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-93edd300f3e0d705dfe5647471697e62c1698695f56c65e766221ef0985d859cL85-R85), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-93edd300f3e0d705dfe5647471697e62c1698695f56c65e766221ef0985d859cL239-R240))
* Simplify the logic of getting `horizontalSize` and `verticalSize` from `size` prop in `index.tsx` and use them as `rowGap` and `columnGap` values or class names ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L34-R40), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L65-R64), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948R80-R81))
* Remove `marginDirection` prop and variable from `index.tsx` and `Item.tsx` as they are no longer needed ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L97-L98), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L114), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a7c0ed1303102439a0949fae071f0b7aa0c4a13694ba55cd513e613c5f3673a2L9), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a7c0ed1303102439a0949fae071f0b7aa0c4a13694ba55cd513e613c5f3673a2L19))
* Add custom token properties for preset sizes of `rowGap` and `columnGap` and use `mergeToken` function to merge them with default values in `style/index.tsx` ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a33a8c350fbb146cd17cd02b03389dc4472d7a18937f7df0d5006286a1ecee93L2-R2), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a33a8c350fbb146cd17cd02b03389dc4472d7a18937f7df0d5006286a1ecee93L11-R13), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a33a8c350fbb146cd17cd02b03389dc4472d7a18937f7df0d5006286a1ecee93L48-R90))
* Add `genSpaceGapStyle` function to generate style rules for preset sizes of `rowGap` and `columnGap` in `style/index.tsx` ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-a33a8c350fbb146cd17cd02b03389dc4472d7a18937f7df0d5006286a1ecee93L48-R90))
* Add blank lines after imports in `index.test.tsx` and `style/compact.tsx` to follow code style convention ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-93edd300f3e0d705dfe5647471697e62c1698695f56c65e766221ef0985d859cR3), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-fc7c7ef13a00b084828f937f3b3189c8b31fec4ad8ece5dcf2f25aaf4619ae13R2))
* Reorder imports to have React as the first import in `demo/base.tsx` and `index.tsx` to follow code style convention ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-ac6caa3882622ab5442cd6d772e217963e9435a07b597e346e73e722cf1c3b50L1-R2), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L3-R13))
* Remove unnecessary line breaks in `style` prop in `index.tsx` to follow code style convention ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-d42c3072a5016964aad21722cb773b77e6fc5824adf56dc16efaecff82aab948L154-R155))
* Assign query results to `items` variable for readability in `index.test.tsx` ([link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-93edd300f3e0d705dfe5647471697e62c1698695f56c65e766221ef0985d859cL69-R72), [link](https://github.com/ant-design/ant-design/pull/44598/files?diff=unified&w=0#diff-93edd300f3e0d705dfe5647471697e62c1698695f56c65e766221ef0985d859cL98-R98))

### CODE REVIEW

File: components/space/Item.tsx [hunk 1]
```diff
@@ -16,27 +17,28 @@ const Item: React.FC<ItemProps> = ({
...
+      if (index < latestIndex && horizontalSize) {
         style = { marginBottom: horizontalSize / (split ? 2 : 1) };
       }
     } else {
-      style = {
-        ...(index < latestIndex && { [marginDirection]: horizontalSize / (split ? 2 : 1) }),
-        ...(wrap && { paddingBottom: verticalSize }),
-      };
+      if (index < latestIndex && horizontalSize) {
+        style = { marginInlineStart: horizontalSize / (split ? 2 : 1) };
```
Comment (li-jia-nan) [inline components/space/Item.tsx:37]: 顺便把方向属性换成CSS逻辑属性了

File: components/space/index.tsx [hunk 2]
```diff
@@ -31,15 +32,19 @@ export interface SpaceProps extends React.HTMLAttributes<HTMLDivElement> {
...
+    return false;
+  }
+  return typeof size === 'string' && ['small', 'middle', 'large'].includes(size);
 };
 
-function getNumberSize(size: SpaceSize) {
-  return typeof size === 'string' ? spaceSize[size] : size || 0;
-}
+const isValidNumber = (size: SpaceSize): size is number => {
+  if (!size) {
```
Comment (zombieJ) [inline components/space/index.tsx:43]: 这个 if 没有必要，会把 `size=0` 变成 false

File: components/space/index.tsx [hunk 3]
```diff
@@ -137,25 +140,25 @@ const Space = React.forwardRef<HTMLDivElement, SpaceProps>((props, ref) => {
     gapStyle.flexWrap = 'wrap';
 
     // Patch for gap not support
-    if (!supportFlexGap) {
+    if (!supportFlexGap && isValidNumber(verticalSize)) {
```
Comment (zombieJ) [inline components/space/index.tsx:143]: `isValidNumber` 出现了很多次，直接在 render 里写一个 `isValidXXXXNumber` 变量抽一下。

File: components/space/index.tsx (hunk 2 above)
Comment (li-jia-nan) [reply components/space/index.tsx:43]: 这个是故意的，因为 `gap` 属性在 CSS 中的默认值就是0，所以如果用户传0进来的话，我们可以直接忽略

File: components/space/index.tsx (hunk 3 above)
Comment (li-jia-nan) [reply components/space/index.tsx]: done~

File: components/space/utils.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import type { SpaceSize } from '.';
+import type { SizeType } from '../config-provider/SizeContext';
+
+export const isPresetSize = (size: SpaceSize): size is SizeType => {
+  if (!size) {
```
Comment (zombieJ) [inline components/space/utils.ts:5]: 这边也不用了，下面的代码已经覆盖到了

Comment (li-jia-nan) [reply components/space/utils.ts:5]: done~

**Review by zombieJ — APPROVED** [review]

**Comment by zombieJ** [comment]
CI 挂了

截屏2023-09-03 15 35 06

**Review by zombieJ — CHANGES_REQUESTED** [review]
看看为什么镜像比较挂了

**Comment by li-jia-nan** [comment]
> 看看为什么镜像比较挂了

done~

File: components/space/Item.tsx [hunk 5]
```diff
@@ -34,7 +34,7 @@ const Item: React.FC<ItemProps> = ({
       }
     } else {
       style = {
-        ...(index < latestIndex && { [marginDirection]: horizontalSize / (split ? 2 : 1) }),
+        ...(index < latestIndex && { marginInlineEnd: horizontalSize / (split ? 2 : 1) }),
```
Comment (zombieJ) [inline components/space/Item.tsx:37]: 突然想起来不对，css logic props 出现的比 flex 晚。如果支持了 css logic prop 则必然支持 flex。不支持 flex 的，这个兜底也会挂。

Comment (zombieJ) [reply components/space/Item.tsx:37]: 到时候加个 comment，免得下次又不小心改了

Comment (li-jia-nan) [reply components/space/Item.tsx]: done~

File: components/space/index.tsx [hunk 6]
```diff
@@ -60,30 +53,54 @@ const Space = React.forwardRef<HTMLDivElement, SpaceProps>((props, ref) => {
     ...otherProps
   } = props;
 
-  const supportFlexGap = useFlexGapSupport();
+  const { token } = theme.useToken();
```
Comment (zombieJ) [inline components/space/index.tsx:56]: 内部有个自己的 useToken，不要用 theme 的

File: components/space/index.tsx [hunk 7]
```diff
@@ -60,30 +53,54 @@ const Space = React.forwardRef<HTMLDivElement, SpaceProps>((props, ref) => {
...
+  }, [horizontalSize]);
+
+  const realVerticalSize = React.useMemo<number>(() => {
+    if (isPresetSize(verticalSize)) {
+      return spaceSizeMap[verticalSize!];
+    }
+    if (isValidGapNumber(verticalSize)) {
+      return verticalSize;
+    }
+    return 0;
```
Comment (zombieJ) [inline components/space/index.tsx:83]: 这个和上面几乎一模一样，直接抽成一个方法

File: components/space/index.tsx (hunk 6 above)
Comment (li-jia-nan) [reply components/space/index.tsx]: done~

File: components/space/__tests__/index.test.tsx [hunk 8]
```diff
@@ -231,25 +213,29 @@ describe('Space', () => {
...
+    expect(
+      container.querySelector<HTMLDivElement>('.ant-space-item')?.getAttribute('style'),
+    ).toEqual('margin-right: 8px; color: red;');
+  });
 
-    expect(container.querySelector('.ant-space-item')?.getAttribute('style')).toEqual(
-      'margin-right: 8px; color: red;',
-    );
+  it('should isPresetSize & isValidGapNumber work', () => {
+    const r1 = isPresetSize(0);
```
Comment (zombieJ) [inline components/space/__tests__/index.test.tsx:232]: 不要测试内部函数，这样会误导覆盖率

File: components/space/utils.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import type { SpaceSize } from '.';
+import type { SizeType } from '../config-provider/SizeContext';
+
+export function isPresetSize(size: SpaceSize): size is SizeType {
+  return ['small', 'middle', 'large'].includes(size as string);
+}
+
+export function isValidGapNumber(size: SpaceSize): size is number {
```
Comment (zombieJ) [inline components/space/utils.ts:8]: 这个方法外面没有用到，不用 export

File: components/space/__tests__/index.test.tsx (hunk 8 above)
Comment (li-jia-nan) [reply components/space/__tests__/index.test.tsx]: 删掉了

File: components/space/utils.ts (hunk 9 above)
Comment (li-jia-nan) [reply components/space/utils.ts]: 删掉了

**Review by zombieJ — APPROVED** [review]

**Comment by zombieJ** [comment]
紧凑模式也对了，不过这个看起来很像是个 breaking change

截屏2023-09-04 19 56 30

**Pull request merged**